relay-server/src/http.rs

```diff
@@ -0,0 +1,299 @@
+    where
+        E: From<reqwest::Error> + From<ActixError>,
+    {
+        match self {
+            RequestBuilder::Actix(mut builder) => Ok(Request::Actix(builder.finish()?)),
+            RequestBuilder::Reqwest {
+                builder,
+                http_encoding: _,
+            } => Ok(Request::Reqwest(builder.build()?)),
+        }
+    }
+
+    /// Add a new header, not replacing existing ones.
+    pub fn header(&mut self, key: &str, value: &[u8]) -> &mut Self {
+        match self {
+            RequestBuilder::Actix(builder) => {
+                builder.header(key, value);
+            }
+            RequestBuilder::Reqwest { builder, .. } => {
+                take_mut::take(builder, |b| b.header(key, value))
+            }
+        }
+
+        self
+    }
+
+    pub fn body<E>(self, body: Binary) -> Result<Request, E>
+    where
+        E: From<reqwest::Error> + From<ActixError> + From<io::Error>,
+    {
+        match self {
+            RequestBuilder::Actix(mut builder) => Ok(Request::Actix(builder.body(body)?)),
+            RequestBuilder::Reqwest {
+                mut builder,
+                http_encoding,
+            } => {
+                let body = match http_encoding {
+                    HttpEncoding::Identity => {
+                        builder = builder.header("Content-Encoding", "identity");
+                        body.as_ref().to_vec()
+                    }
+                    HttpEncoding::Deflate => {
+                        builder = builder.header("Content-Encoding", "deflate");
+                        let mut encoder = ZlibEncoder::new(Vec::new(), Compression::default());
+                        encoder.write_all(body.as_ref())?;
+                        encoder.finish().unwrap()
+                    }
+                    HttpEncoding::Gzip => {
+                        builder = builder.header("Content-Encoding", "gzip");
+                        let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
+                        encoder.write_all(body.as_ref())?;
+                        encoder.finish().unwrap()
+                    }
+                    HttpEncoding::Br => {
+                        builder = builder.header("Content-Encoding", "br");
+                        let mut encoder = BrotliEncoder::new(Vec::new(), 5);
+                        encoder.write_all(body.as_ref())?;
+                        encoder.finish().unwrap()
+                    }
+                };
+
+                RequestBuilder::Reqwest {
+                    builder: builder.body(body),
+                    http_encoding,
+                }
+                .finish()
+            }
+        }
+    }
+
+    pub fn disable_decompress(&mut self) -> &mut Self {
+        match self {
+            RequestBuilder::Actix(builder) => {
+                builder.disable_decompress();
+            }
+            RequestBuilder::Reqwest { .. } => {
+                // can only be set on client TODO
```

Comment:
This needs to be resolved.

Reply:
This stuff was only necessary because of some actix-web behavior that made no sense. It appears to be unnecessary for reqwest.

I moved the actix-specific workarounds into the forward endpoint such that the abstraction does not have to suffer from this.

Reply:
Actix automatically decompresses all responses that come with `content-encoding`. In the forward endpoint, we do not want this, as we want to pipe the compressed response through. This also plays into default headers.

In all cases other the forward endpoint, we need to decompress all compressed responses. Only the forward endpoint should leave it compressed.

Reply:
I need to take a closer look at this. Apparently the tests have really weak assertions for this part.